### PR TITLE
store raft log/raft sate/apply state in raft cf

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
store raft log/raft sate/apply state in raft cf

### CODE REVIEW

File: src/raftstore/store/engine.rs [hunk 1]
```diff
@@ -73,6 +73,20 @@ pub trait Peekable {
         Ok(Some(m))
     }
 
+    fn get_msg_cf<M>(&self, cf: &str, key: &[u8]) -> Result<Option<M>>
```
Comment (siddontang) [inline src/raftstore/store/engine.rs]: please add test for all new added functions. 

Btw, I think we can use this function to simplify get_msg.

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -433,13 +447,22 @@ impl PeerStorage {
             [(keys::region_raft_prefix(region_id), keys::region_raft_prefix(region_id + 1)),
              (keys::region_meta_prefix(region_id), keys::region_meta_prefix(region_id + 1))];
 
+        let handle = try!(rocksdb::get_cf_handle(&self.engine, CF_RAFT));
         for &(ref start_key, ref end_key) in &ranges {
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: you can use CF_RAFT for raft range, and default for meta range now.

File: src/raftstore/store/worker/compact.rs [hunk 3]
```diff
@@ -65,17 +67,18 @@ impl Runner {
...
+        if let Some((k, _)) = box_try!(task.engine.seek_cf(CF_RAFT, &start_key)) {
             first_idx = box_try!(keys::raft_log_index(&k));
         }
         if first_idx >= task.compact_idx {
             info!("no need to compact");
             return Ok(0);
         }
         let wb = WriteBatch::new();
+        let handle = box_try!(rocksdb::get_cf_handle(&task.engine, CF_RAFT));
         for idx in first_idx..task.compact_idx {
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: I think we can use `delete_in_range_cf` here.

Comment (ngaut) [inline src/raftstore/store/worker/compact.rs]: It should be in another PR, we still need lots of tests.

File: tests/raftstore/test_compact_log.rs [hunk 4]
```diff
@@ -28,7 +29,7 @@ fn test_compact_log<T: Simulator>(cluster: &mut Cluster<T>) {
 
     for (&id, engine) in &cluster.engines {
         let mut state: RaftApplyState =
-            engine.get_msg(&keys::apply_state_key(1)).unwrap().unwrap_or_default();
+            engine.get_msg_cf(CF_RAFT, &keys::apply_state_key(1)).unwrap().unwrap_or_default();
```
Comment (siddontang) [inline tests/raftstore/test_compact_log.rs]: can we extract a must_get_apply_state function?

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/peer_storage.rs [hunk 5]
```diff
@@ -363,7 +363,7 @@ impl PeerStorage {
 
         // Delete any previously appended log entries which never committed.
         for i in (last_index + 1)..(prev_last_index + 1) {
-            try!(ctx.wb.delete(&keys::raft_log_key(self.get_region_id(), i)));
+            try!(ctx.wb.delete_cf(*handle, &keys::raft_log_key(self.get_region_id(), i)));
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: seem that we don't have any test here?

**Comment by zhangjinpeng1987** [comment]
@BusyJay @siddontang @disksing PTAL

File: src/raftstore/store/peer_storage.rs (hunk 5 above)
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: any test here? 

/cc @BusyJay

Comment (zhangjinpeng1987) [inline src/raftstore/store/peer_storage.rs]: Is this just for safety? Uncommited logs are store in unstable store, isn't it?

File: src/raftstore/store/peer_storage.rs [hunk 6]
```diff
@@ -340,16 +351,19 @@ impl PeerStorage {
...
         }
 
+        let handle = try!(rocksdb::get_cf_handle(&self.engine, CF_RAFT));
         for entry in entries {
-            try!(ctx.wb.put_msg(&keys::raft_log_key(self.get_region_id(), entry.get_index()),
-                                entry));
+            let value = try!(entry.write_to_bytes());
+            try!(ctx.wb.put_cf(*handle,
+                               &keys::raft_log_key(self.get_region_id(), entry.get_index()),
+                               &value));
```
Comment (disksing) [inline src/raftstore/store/peer_storage.rs]: I suppose it's better to use `put_msg_cf`.

File: src/raftstore/store/peer_storage.rs (hunk 5 above)
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: we have no unstable store, all logs are saved in rocksdb. 

@BusyJay

File: tests/raftstore/test_compact_log.rs [hunk 7]
```diff
@@ -14,21 +14,30 @@
...
 
 use super::util::*;
 use super::cluster::{Cluster, Simulator};
 use super::node::new_node_cluster;
 use super::server::new_server_cluster;
 
+fn must_get_msg_cf<M>(engine: &DB, cf: &str, key: &[u8]) -> M
+    where M: protobuf::Message + protobuf::MessageStatic
+{
+    engine.get_msg_cf(cf, key).unwrap().unwrap_or_default()
```
Comment (siddontang) [inline tests/raftstore/test_compact_log.rs]: IMO, must here means the key must exist, but unwrap_or_default means key maybe not exist.

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay

**Comment by disksing** [comment]
LGTM

File: src/raftstore/store/peer_storage.rs (hunk 5 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Yes, all logs are saved in rocksdb and unstable vector. This loop is just for cleanup. If a region is active and keep appending log, this loop can be removed.

File: src/raftstore/store/peer.rs [hunk 8]
```diff
@@ -987,11 +992,15 @@ struct ExecContext<'a> {
     pub apply_state: RaftApplyState,
     pub wb: WriteBatch,
     pub req: &'a RaftCmdRequest,
+    pub engine: Arc<DB>,
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: I prefer use the actual cf here.

Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: How about if we need is not only raft cf?

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Then we can abstract a trait.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

**Comment by BusyJay** [comment]
LGTM
